.github/workflows/docker-image.yml

```diff
@@ -0,0 +1,84 @@
+name: Docker Image
+
+on:
+  push:
+    branches: [ main ]
+  pull_request:
+    branches: [ main ]
+
+jobs:
+
+  build-login-push-test:
+
+    runs-on: ubuntu-latest
+
+    steps:
+     - uses: actions/checkout@v2
+     - 
+       name: 1 Build the Docker image (2 stages)
+       run: docker build . --build-arg PYPI_INDEX_URL=${{ secrets.PYPI_INDEX_URL }} --file Dockerfile --tag ${{ secrets.PYPI_HOST }}/default-docker-local/python-wallet-api:latest
+     - 
+       name: 2 Docker Login (JFrog repo)
+       uses: docker/login-action@v1
+       with:
+          registry: ${{ secrets.PYPI_HOST }}
+          username: ${{ secrets.PYPI_USERNAME }}
+          password: ${{ secrets.PYPI_PASSWORD }}
+     - 
+         name: 3 Push the image (JFrog repo)
+         run: docker push ${{ secrets.PYPI_HOST }}/default-docker-local/python-wallet-api:latest
+     -
+         name: 4 Do a test container run
+         run: docker run --name python-wallet-api ${{ secrets.PYPI_HOST }}/default-docker-local/python-wallet-api:latest & 
+     -    
+         name: 5 Do docker ps
+         run: docker ps
+     -    
+         name: 6 Stop the container 
+         run: docker stop python-wallet-api
+         
+  # kind-create-kind-cluster:
+  #
+  #  runs-on: ubuntu-latest
+  #  steps:
+  #    - uses: actions/checkout@v2
+  #    - uses: engineerd/setup-kind@v0.5.0
+  #    - name: Testing the cluster
+  #      run: |
+  #        kubectl cluster-info
+  #        kubectl get pods -n kube-system
+  #        echo "current-context:" $(kubectl config current-context)
+  #        echo "environment-kubeconfig:" ${KUBECONFIG}
+  #    - name: Deploy app & service to kinD
+  #      run: |
+  #        kubectl apply -f $HOME/deployment.yml
+  #        kubectl get pod -l app=wallet-api-ojson | jq .status.phase 
+  
+  k3s-create-cluster-then-deploy-app:
+
+      runs-on: ubuntu-latest
+
+      steps:
+      - uses: actions/checkout@v2
+      - uses: nolar/setup-k3d-k3s@v1
+      - 
+       name: Get k3s cluster info
+       run: |
+          kubectl cluster-info
+          kubectl get pods -n kube-system
+          kubectl get nodes
+          echo "current-context:" $(kubectl config current-context)
+          echo "environment-kubeconfig:" ${KUBECONFIG}
+      - 
+       name: Deploy app & service to k3d
+       run: |
+          ls .
+          kubectl apply -f ./deployment.yml
+          kubectl get pods
+          echo "Waiting for pods... 1 min..." && sleep 60
```

Comment:
Sleeping for an arbitrary timeout is not an optimal solution. It would be better to actually check the pod's status in a loop. And limit that loop to a number of iterations whereas each iteration sleeps for a longer period (exponential backoff) until it gives up.

Reply:
What kind of loop you mean?